diff --git a/avs_core/CMakeLists.txt b/avs_core/CMakeLists.txt
index 76e61caa..51f5dcd3 100644
--- a/avs_core/CMakeLists.txt
+++ b/avs_core/CMakeLists.txt
@@ -97,26 +97,26 @@ if (EXISTS ${GHS_FILESYSTEM_INCLUDE_DIR})
     target_include_directories("AvsCore" SYSTEM PRIVATE ${GHS_FILESYSTEM_INCLUDE_DIR})
 endif()
 
-# stdc++fs was mainlined into stdc++ in GCC 9, but GCC 8 can build it too
-if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9)
-    set(FSLIB "stdc++fs")
-endif()
-
 if (CMAKE_SYSTEM_NAME STREQUAL "Haiku")
     set(SYSLIB "root")
-    set(SYSLIBS "-lroot")
+elseif(MSVC OR MINGW)
+    set(SYSLIB "uuid" "winmm" "vfw32" "msacm32" "gdi32" "user32" "advapi32" "ole32" "imagehlp")
 else()
-    set(SYSLIB "pthread" "dl")
-    set(SYSLIBS "-lpthread -ldl")
+    set(SYSLIB "pthread" "dl" "m")
 endif()
 
-# Per-OS link dependencies
-if (MSVC OR MINGW)
-  target_link_libraries("AvsCore" "uuid" "winmm" "vfw32" "msacm32" "gdi32" "user32" "advapi32" "ole32" "imagehlp")
-else()
-  target_link_libraries("AvsCore" "${SYSLIB}" "${FSLIB}")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    list(APPEND SYSLIB "stdc++")
+    # stdc++fs was mainlined into stdc++ in GCC 9, but GCC 8 can build it too
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9)
+        list(APPEND SYSLIB "stdc++fs")
+    endif()
+elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    list(APPEND SYSLIBFLAGS "-stdlib=libc++")
 endif()
 
+target_link_libraries("AvsCore" ${SYSLIB})
+
 if (MSVC_IDE)
   # Copy output to a common folder for easy deployment
   add_custom_command(
@@ -165,6 +165,10 @@ endif()
 # Generate pkg-config file
 get_target_property(LIB_NAME AvsCore OUTPUT_NAME)
 set(LIBS "-l${LIB_NAME}")
+set(_SYSLIBS_ITEMS ${SYSLIB})
+list(TRANSFORM _SYSLIBS_ITEMS PREPEND "-l")
+list(APPEND _SYSLIBS_ITEMS ${SYSLIBFLAGS})
+list(JOIN _SYSLIBS_ITEMS " " SYSLIBS)
 CONFIGURE_FILE("avisynth.pc.in" "avisynth.pc" @ONLY)
 
 # Generate avisynth.conf
diff --git a/avs_core/avisynth.pc.in b/avs_core/avisynth.pc.in
index 6922c996..c5f2bacb 100644
--- a/avs_core/avisynth.pc.in
+++ b/avs_core/avisynth.pc.in
@@ -10,3 +10,4 @@ Version: @PROJECT_VERSION@
 Libs: -L${libdir} @LIBS@
 Libs.private: -L${libdir} @SYSLIBS@
 Cflags: -I${includedir}
+Cflags.private: -DAVS_STATIC_LIB
