diff --git a/avs_core/CMakeLists.txt b/avs_core/CMakeLists.txt
index dc5000f6..23abcf8e 100644
--- a/avs_core/CMakeLists.txt
+++ b/avs_core/CMakeLists.txt
@@ -101,31 +101,27 @@ if (EXISTS ${GHS_FILESYSTEM_INCLUDE_DIR})
     target_include_directories("AvsCore" SYSTEM PRIVATE ${GHS_FILESYSTEM_INCLUDE_DIR})
 endif()
 
-# stdc++fs was mainlined into stdc++ in GCC 9, but GCC 8 can build it too
-if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9)
-    set(FSLIB "stdc++fs")
-endif()
-
 if (CMAKE_SYSTEM_NAME STREQUAL "Haiku")
     set(SYSLIB "root")
-    set(SYSLIBS "-lroot")
-else()
-if (CMAKE_SYSTEM_NAME STREQUAL "OpenBSD")
-    set(SYSLIB "pthread")
-    set(SYSLIBS "-lpthread")
+elseif (CMAKE_SYSTEM_NAME STREQUAL "OpenBSD")
+    set(SYSLIB "m" "pthread")
+elseif(MSVC OR MINGW)
+    set(SYSLIB "uuid" "winmm" "vfw32" "msacm32" "gdi32" "user32" "advapi32" "ole32" "imagehlp")
 else()
-    set(SYSLIB "pthread" "dl")
-    set(SYSLIBS "-lpthread -ldl")
-endif()
+    set(SYSLIB "m" "pthread" "dl")
 endif()
 
-# Per-OS link dependencies
-if (MSVC OR MINGW)
-  target_link_libraries("AvsCore" "uuid" "winmm" "vfw32" "msacm32" "gdi32" "user32" "advapi32" "ole32" "imagehlp")
-else()
-  target_link_libraries("AvsCore" "${SYSLIB}" "${FSLIB}")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    list(APPEND SYSLIB "stdc++")
+    # stdc++fs was mainlined into stdc++ in GCC 9, but GCC 8 can build it too
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9)
+        list(APPEND SYSLIB "stdc++fs")
+    endif()
+elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    list(APPEND SYSLIB "c++")
 endif()
 
+target_link_libraries("AvsCore" ${SYSLIB})
 
 if(ENABLE_CUDA)
   if(${CMAKE_VERSION} VERSION_LESS "3.19.5")
@@ -136,6 +132,7 @@ if(ENABLE_CUDA)
       #target_link_libraries ("AvsCore" ${CUDA_LIBRARIES})
       target_link_libraries ("AvsCore" ${CUDA_cudart_static_LIBRARY})
       target_compile_definitions("AvsCore" PUBLIC ENABLE_CUDA)
+      list(APPEND SYSLIB "cudart_static")
     endif()
   else()
     # Fixed in 3.19.5: https://gitlab.kitware.com/cmake/cmake/-/issues/21740
@@ -145,6 +142,7 @@ if(ENABLE_CUDA)
       target_include_directories("AvsCore" PRIVATE ${CUDAToolkit_INCLUDE_DIRS})
       target_link_libraries ("AvsCore" CUDA::cudart_static)
       target_compile_definitions("AvsCore" PUBLIC ENABLE_CUDA)
+      list(APPEND SYSLIB "cudart_static")
     endif()
   endif()
 endif()
@@ -197,6 +195,9 @@ endif()
 # Generate pkg-config file
 get_target_property(LIB_NAME AvsCore OUTPUT_NAME)
 set(LIBS "-l${LIB_NAME}")
+set(_SYSLIBS_ITEMS ${SYSLIB})
+list(TRANSFORM _SYSLIBS_ITEMS PREPEND "-l")
+list(JOIN _SYSLIBS_ITEMS " " SYSLIBS)
 CONFIGURE_FILE("avisynth.pc.in" "avisynth.pc" @ONLY)
 
 # Generate avisynth.conf
diff --git a/avs_core/avisynth.pc.in b/avs_core/avisynth.pc.in
index 6922c996..becb0e43 100644
--- a/avs_core/avisynth.pc.in
+++ b/avs_core/avisynth.pc.in
@@ -8,5 +8,6 @@ Description: A powerful nonlinear scripting language for video.
 Version: @PROJECT_VERSION@
 
 Libs: -L${libdir} @LIBS@
-Libs.private: -L${libdir} @SYSLIBS@
+Libs.private: @SYSLIBS@
 Cflags: -I${includedir}
+Cflags.private: -DAVS_STATIC_LIB
