diff --git a/avs_core/avisynth.pc.in b/avs_core/avisynth.pc.in
index 6922c996..5c3b920b 100644
--- a/avs_core/avisynth.pc.in
+++ b/avs_core/avisynth.pc.in
@@ -1,7 +1,7 @@
 prefix=@CMAKE_INSTALL_PREFIX@
 exec_prefix=@CMAKE_INSTALL_PREFIX@
 libdir=${exec_prefix}/@CMAKE_INSTALL_LIBDIR@
-includedir=${exec_prefix}/@CMAKE_INSTALL_INCLUDEDIR@/avisynth
+includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@/avisynth
 
 Name: AviSynth+
 Description: A powerful nonlinear scripting language for video.
